Trust Spawn/Despawn/Death/Teamwork & Error Messages

zach2good opened the pull request:
Ended up completing this feature once I tracked down the required packet and the message offsets.

**Done:**
messageCombat and trust messages

Swapping out hardcoded values for spell table entries.

Swap out hard-coded trust ID values for tabled spellID values in these:

Table the system_messages used in trust.lua

Maybe rename getSeekingParty to isSeekingParty

Add timer for last time a player joined a party or when the party was formed, for the last trust error message

<!-- place 'x' mark between square [] brackets to affirm: -->
**_I affirm:_**
- [x] that I agree to Project Topaz's [Limited Contributor License Agreement](http://project-topaz.com/blob/release/CONTRIBUTOR_AGREEMENT.md), as written on this date
- [x] that I've _tested my code_ since the last commit in the PR, and will test after any later commits

zach2good committed:
Quick investigation into Trust Messages

zach2good committed:
Figure out offsets between trusts

zach2good committed:
Hook up other nation trusts and find offset for special moves

TeoTwawki commented:
might be worth renaming that packet/message while its still young: 
https://github.com/project-topaz/topaz/search?q=CMessageDebugPacket&unscoped_q=CMessageDebugPacket

zach2good committed:
Rename CMessageDebugPacket -> CMessageFinishPacket

zach2good committed:
Naji <3 Ayame

zach2good committed:
More synergy messages

zach2good committed:
CMessageFinishPacket -> CMessageCombatPacket

zach2good committed:
Add Synergy message slot 5, use spell tables for everything, cleanup

zach2good committed:
Rename binding to: messageCombat

zach2good committed:
Fix Kupipi spells typo

zach2good committed:
Send message to party members

zach2good committed:
Add trust-specific system error messages and conditions

zach2good committed:
Remove uneeded require for spell_data

zach2good committed:
Table trust system messages

zach2good committed:
Replace nation trust duplicate checks with spell table values

zach2good committed:
getSeekingParty -> isSeekingParty

zach2good committed:
Add msgsystem enums to core

zach2good committed:
Start work on binding for getPartyLastMemberJoinedTime

TeoTwawki commented on scripts/globals/msg.lua:
```diff
@@ -311,3 +311,17 @@ tpz.msg.basic =
+------------------------------------
+-- Message System
+------------------------------------
+
```
since you renamed the paclket, maybe rename the table to match its new name?

TeoTwawki replied:
OH! Ima dunce these are yellow system3 type aren't they, i mistook this as an enum for the combatMessage packet you did.

TeoTwawki commented on scripts/globals/spells/trust/ajido-marujido.lua:
```diff
@@ -27,10 +27,20 @@ function onMobSpawn(mob)
     mob:addSimpleGambit(ai.t.TARGET, ai.c.NOT_SC_AVAILABLE, 0,
                         ai.r.MA, ai.s.HIGHEST, tpz.magic.spellFamily.NONE, 30)
+
+    tpz.trust.synergyMessage(mob, {
```
I expect using the term "synergy" will be confusing once we have synergy craft skill work started.

zach2good committed:
Rename some things

zach2good committed:
More renaming

zach2good committed:
Missed one

zach2good committed:
Fix member added time and test it works

zach2good committed:
Add documentation

Omnione commented on src/map/packets/message_system.h:
```diff
@@ -26,6 +26,16 @@
+    MSGSYSTEM_TRUST_NO_ENMITY                = 300,   // You cannot use Trust magic while having gained enmity.
+    MSGSYSTEM_TRUST_SOLO_OR_LEADER           = 301,   // You cannot use Trust magic unless you are solo or the party leader.
+};
+
```
Does this need to be here rather than msg.lua?, im only asking as we usually put these messages in msg.lua and call from that, and it seems that you allready call tpz.msg for messages.

zach2good replied:
It doesn't _need_ to be here, but this is following the pattern from basic messages since everything has to flow through core to be able to have the right packet sent out

Omnione replied:
Ah ok, got ya, ok carry on lol

Omnione replied:
and wow totaly missed that MSGSYSTEM_MSGSYSTEM lol, glad you caught that

Omnione reviewed (commented):
Other than the messages, this looks really good.

zach2good committed:
Fix duplication in name

zach2good committed:
Alignment

ibm2431 commented on scripts/globals/trust.lua:
```diff
@@ -86,3 +115,43 @@ tpz.trust.spawn = function(caster, spell)
+
+tpz.trust.message = function(mob, id)
+    local master = mob:getMaster()
+    local offset = (mob:getTrustID() - 896) * 100
+    local party = master:getParty()
+
+    for _, member in ipairs(party) do
+        member:messageCombat(mob, offset + id, 0, 711)
```
I worry about starting message ID shifting. I don't know the last time it's happened on the low end, but I don't know if this trust global will be on the minds of people running text ID tools should it happen.

Would it be possible to define a very-first-starting-trust message ID in `msg`, and then calculate offsets based on that?
`local message_id = tpz.msg.combat.GLOBAL_TRUST_OFFSET + ((trust_id - x) * 100)`
The global would need to be adjusted if the _size_ of trust message blocks change, but they'd be a little more resistant if SE decides to throw a couple random messages before the trust block.

Variable naming could be a bit clearer - what's being passed in isn't a message ID, but a message offset. Complication in that you're also calculating a _trust_ offset. Maybe `trust_offset` + `message_offset`.

TeoTwawki replied:
if its a per zone thing (as in, its in the string dats of each zone, even if they all have the same ID now) it should really be in the IDs.lua anyway just for this reason. if its not in the zones dat, we generally won't need to worry about it shifting. (I'm still in favor of offsets where we can, vs enuming all the stuff and things, just FYI))

I most zones we have several IDs that are generally the same in every zone - we still IDs.lua them because they will eventually get shifted,
![image](https://user-images.githubusercontent.com/6871475/87230484-5e15bd00-c37e-11ea-9695-ff97cd618c2d.png)

ibm2431 commented on scripts/globals/trust.lua:
```diff
@@ -86,3 +115,43 @@ tpz.trust.spawn = function(caster, spell)
+        end
+    end
+
+    if table.getn(messages) > 0 then
+        tpz.trust.message(mob, messages[math.random(#messages)])
```
Are we certain the message selected is random? There's no priority system? I've done a lot of partying with ROV versions of The Princesses, and I remember them _always_ commenting about Iroha.

I don't know how far your retail character is, so I'll tag @project-topaz/researcher :
1. Summon Iroha II
2. Summon The Princesses II
3. See if The Princesses ever comment about each other versus Iroha

zach2good replied:
_Seems_ random...

TeoTwawki replied:
~~I believe there is a priority system of sorts based on what order they were summoned in on retail, but someone will need to confirm that~~ _Vidya says I'm wrong_

ibm2431 commented on scripts/globals/trust.lua:
```diff
@@ -86,3 +115,43 @@ tpz.trust.spawn = function(caster, spell)
+    end
+end
+
+-- For debugging and lining up teamwork messages
+tpz.trust.dumpMessages = function(mob)
+    for i=0, 20 do
+        tpz.trust.message(mob, i)
+    end
```
👍

ibm2431 commented on src/map/lua/lua_baseentity.cpp:
```diff
@@ -533,6 +534,39 @@ inline int32 CLuaBaseEntity::messageSystem(lua_State* L)
+
+    CCharEntity* PChar = (CCharEntity*)m_PBaseEntity;
+
+    CBaseEntity* PTarget;
```
"Target" or "Speaker"?

ibm2431 commented on src/map/packets/message_system.h:
```diff
@@ -26,6 +26,16 @@
 
 #include "basic.h"
 
+enum MSGSYSTEM_ID : uint16
+{
+    MSGSYSTEM_TRUST_NO_SEEKING_PARTY         = 296,   // You cannot use Trust magic while seeking a party.
+    MSGSYSTEM_TRUST_DELAY_NEW_PARTY_MEMBER   = 297,   // While inviting a party member, you must wait a while before using Trust magic.
+    MSGSYSTEM_TRUST_MAXIMUM_NUMBER           = 298,   // You have called forth your maximum number of alter egos.
+    MSGSYSTEM_TRUST_ALREADY_CALLED           = 299,   // That alter ego has already been called forth.
+    MSGSYSTEM_TRUST_NO_ENMITY                = 300,   // You cannot use Trust magic while having gained enmity.
+    MSGSYSTEM_TRUST_SOLO_OR_LEADER           = 301,   // You cannot use Trust magic unless you are solo or the party leader.
+};
```
You're not using these enums in core~!

zach2good replied:
I've decided to remove them (for now), because I don't want to chase down all of the other hard coded uses of messageSystem...

ibm2431 commented on scripts/globals/trust.lua:
```diff
@@ -28,14 +38,27 @@ tpz.trust.canCast = function(caster, spell, not_allowed_trust_ids)
+        return -1
+    end
 
+    -- Block summoning trusts if someone recently joined party (120s)
```
I don't remember, but are we already blocking party invites to players if trusts are present? (Either a party leader using trusts sending an invite to a free person, or a player-only party sending an invite to someone with trusts out)

Not a requirement for this PR if we're not; just making sure we don't forget any loose threads~

zach2good replied:
Not really, on the core end everything is pretty much unlimited and the only blocking happens on the lua side. The way trusts are handled in and around a player/mixed party is all sorts of wrong. That'll have to be a whole research and dev leg on its own before release :(

ibm2431 reviewed (commented):
Thanks, Zach!

zach2good committed:
Review items:
- Global trust message offset
- Clearer naming and offset use in trust.message
- PTarget -> PSpeaker
- Remove unused system message enums

zach2good committed:
Restore trampled excenmille and naji balance mods

zach2good committed:
Use magic require instead of spell_data

ibm2431 reviewed (approved):
Looks good to me! Thanks, Zach!

zach2good committed:
Rebase and cleanup
